File: Data Structures & Algorithms/valid-parenthesis-string/submission-0.java
class Solution {
    boolean rec(char[]ch,int i,int left){
        if(i==ch.length){
           return left==0;
        }
        if(left<0)return false;
        if(ch[i]=='(')left++;
        else if(ch[i]==')')left--;
        else{
            // if(rec(ch,i+1,left))return true;
            if(rec(ch,i+1,left-1))return true;
            if(rec(ch,i+1,left+1))return true;
        }
        if(rec(ch,i+1,left))return true;
        return false;
    }
    public boolean checkValidString(String s) {
        char ch[]=s.toCharArray();
        return rec(ch,0,0);
    }
}

File: Data Structures & Algorithms/coin-change/submission-0.py
from functools import cache
class Solution:
    
    def coinChange(self, coins: List[int], amount: int) -> int:
        mx = 10000000007
        coins = sorted(coins)[::-1]
        @cache
        def get(i, left):
            if left == 0:
                return 0
            if i == len(coins):
                return mx
            if coins[i] > left:
                return get(i + 1, left)
            return min(get(i + 1, left), get(i, left - coins[i]) + 1)
        x = get(0, amount) 
        return x if x < mx else -1

            

        